Definición del modelo

In [1]:
from tensorflow import keras
from tensorflow.keras import layers
inputs = keras.Input(shape=(28, 28, 1))
x = layers.Conv2D(filters=32, kernel_size=3, activation="relu")(inputs)
x = layers.MaxPooling2D(pool_size=2)(x)
x = layers.Conv2D(filters=64, kernel_size=3, activation="relu")(x)
x = layers.MaxPooling2D(pool_size=2)(x)
x = layers.Conv2D(filters=128, kernel_size=3, activation="relu")(x)
x = layers.Flatten()(x)
outputs = layers.Dense(10, activation="softmax")(x)
model = keras.Model(inputs=inputs, outputs=outputs)

Verificación de la arquitectura de la red neuronal convolucional

In [2]:
model.summary()

Model: "model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_1 (InputLayer)        [(None, 28, 28, 1)]       0         
                                                                 
 conv2d (Conv2D)             (None, 26, 26, 32)        320       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 13, 13, 32)       0         
 )                                                               
                                                                 
 conv2d_1 (Conv2D)           (None, 11, 11, 64)        18496     
                                                                 
 max_pooling2d_1 (MaxPooling  (None, 5, 5, 64)         0         
 2D)                                                             
                                                                 
 conv2d_2 (Conv2D)           (None, 3, 3, 128)         73856 

Cargar MNIST, ajuste de valores, definición de hiperparámetros y entrenamiento del modelo

In [4]:
from keras.datasets import mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32') / 255
test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32') / 255

model.compile(optimizer='rmsprop',
      loss='sparse_categorical_crossentropy',
      metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=10, batch_size=32)

Epoch 1/10
1875/1875 [==============================] - 59s 31ms/step - loss: 0.0260 - accuracy: 0.9923
Epoch 2/10
1875/1875 [==============================] - 57s 30ms/step - loss: 0.0217 - accuracy: 0.9939
Epoch 3/10
1875/1875 [==============================] - 57s 31ms/step - loss: 0.0185 - accuracy: 0.9944
Epoch 4/10
1875/1875 [==============================] - 56s 30ms/step - loss: 0.0144 - accuracy: 0.9959
Epoch 5/10
1875/1875 [==============================] - 59s 31ms/step - loss: 0.0131 - accuracy: 0.9963
Epoch 6/10
1875/1875 [==============================] - 58s 31ms/step - loss: 0.0114 - accuracy: 0.9968
Epoch 7/10
1875/1875 [==============================] - 58s 31ms/step - loss: 0.0103 - accuracy: 0.9973
Epoch 8/10
1875/1875 [==============================] - 58s 31ms/step - loss: 0.0087 - accuracy: 0.9977
Epoch 9/10
1875/1875 [==============================] - 57s 30ms/step - loss: 0.0083 - accuracy: 0.9978
Epoch 10/10
1875/1875 [==============================] - 58s 31m

Verificación del accuracy 

In [ ]:
test_loss, test_acc = model.evaluate(test_images, test_labels)
print('test_loss: ',test_loss)
print('test_acc: ',test_acc)

Mejora la precisión de 0.97 de MNIST en una red neuronal superficial a 0.99 para una red neuronal convolucional ... La perdida es más pequeña pues se reduce de 0.072 a 0.022 

Ejercicios:
----------

1. Trate de reducir los tiempos de ejecución jugando con las pasadas (epoch) y los lotes (batch_size)
2. ¿Que diferencia hace si no hay capas de reducción de dimensionalidad o max-pooling?
3. ¿Cómo se calculan los parámetros en una red neuronal convolucional? Busque por internet y asegurese que coincide con los resultados que muestra la función **summary**